pipelines/pipeline/shared/transforms/validate_csv.py

```diff
@@ -0,0 +1,95 @@
+# this is a "functions that validate dataframes" toolbox
+# will be pieced together in another ticket (delete this text when it gets picked up)
+# more functions can be added as needed
+
+
+from typing import List
+from pathlib import Path
+import pandas as pd
+
+def _read_in_csv_check(csv_path):
+    """
+    given a csv will read in first 5 lines as a check
+    """
+    assert csv_path.name.endswith('.csv'), "Invalid csv_path"
+    
+    try:
+        pd.read_csv(csv_path, nrows=5)
+    except:
+        raise Exception(f"Failed to read in csv - {csv_path}")
+        
+        
+def _correct_columns_exist(csv_path, columns: List[str]):
+    """
+    given a list of columns, make sure those columns are in the dataframe
+    """
+    df = pd.read_csv(csv_path, nrows=0) # reading in columns only
+    
+    assert len(df.columns) == len(columns), f"Number of columns in csv file ({len(df.columns)}) does not match len of expected columns ({len(columns)})"
+    
+    for col in columns:
+        assert col in df.columns, f"Expect column {col} not found in csv file"
+        
+    # number of columns are the same and all expected columns are in df so no
+    # need to check the other way around
+
+
+def _dataframe_has_no_blanks(df):
+    """
+    given a df and a column name confirm that there are
+    no blank entries in the column.
+    """
+    for col in df.columns:
+        # will treat '' & pd.isnull as blanks                 
+        for value in df[col].unique():
+            if type(value) == bool:
+                # bool can only be True of False so not empty
+                continue
+            elif type(value) == str:
+                if not value.strip(): # catches whitespace strings
+                    raise ValueError(f"{col} has blank entries - '{value}' found")
+            else:
+                if pd.isnull(value):
+                    raise ValueError(f"{col} has blank entries - '{value}' found")
+                    
+
+def _dataframe_has_no_duplicates(df):
+    """
+    given a df confirm that there are no duplicate values
+    """
+    assert len(df) == len(df.drop_duplicates()), "Found duplicate rows in the dataframe, failed validation"
+    
+
+def dataframing_slicing(csv_path) -> List[pd.core.frame.DataFrame]:
```

Comment:
so this one gets really interesting 😈 

we want to avoid putting the whole thing in memory at once if we can, so the logic needs to be something like:

- rename if `generated_dataframe_slices` or similar.
- count the rows without using a dataframe
- work out something like a dict of rows tracking the start and end of each slice that we want, so its its a 10020 rows csv, we'd want `{0:, 5000, 5000: 10000, 10000, 100021}`
- iterate that dict and for each one yields a dataframe consisting of _only_ that slice of rows (there should be some way of doing that via pd.read_csv). You'll need to make sure that each one has the correct header row too.

what you;re aiming for is to be able to do something like:

```python
for df in generated_dataframe_slices(csv_path):
      # run checks against this dataframe that is a subset of rows
      # this subset is the only thing pandas has in memory
```

which _looks_ simple but the magic trick here is that because it a generator (yield) at any given time _only_ the one subset of rows is in memory, so it'll work against absolutely **massive** datasets without hitting the pandas resource limits.

_note - cant remember if you can iterate a generator in quite that fashion, you. might need to finagle it, but the ideas sound._